### ISSUE TITLE
[data] data.train_max_samples / val_max_samples draw an unseeded random subset, so checkpoint resume silently continues on different data

### ISSUE THREAD

**Issue by YeonwooSung** [verl-project/verl#7816]
### System Info

Reproduced by reading current main (`1252cc71`).

Affects the RL data path on both trainers (`trainer.use_v1` true and false):
- `verl/utils/dataset/rl_dataset.py` (subset selection)
- `verl/trainer/ppo/v1/trainer_base.py` / `verl/trainer/ppo/ray_trainer.py` (dataset construction, checkpoint load order)
- `verl/trainer/config/data/legacy_data.yaml` (`shuffle`, `seed`, `validation_shuffle` defaults)

The SFT path is not affected (`MultiTurnSFTDataset` defaults `shuffle` to `False` and takes a deterministic `np.arange(max_samples)` slice).

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

`data.train_max_samples` / `data.val_max_samples` subsample the dataset. When shuffling is on, the subset is drawn from a generator that is seeded only if `data.seed` happens to be set:

```python
# verl/utils/dataset/rl_dataset.py:185-192
if self.max_samples > 0 and self.max_samples < total:
    if self.shuffle:
        rngs_args = (self.seed,) if self.seed is not None else ()
        rng = np.random.default_rng(*rngs_args)
        indices = rng.choice(total, size=self.max_samples, replace=False)
    else:
        indices = np.arange(self.max_samples)
    self.dataframe = self.dataframe.select(indices.tolist())
```

The relevant defaults make the unseeded branch the default one:

```yaml
# verl/trainer/config/data/legacy_data.yaml
shuffle: True        # :68
seed: null           # :71
validation_shuffle: False   # :80
```

With `seed: null`, `np.random.default_rng()` seeds itself from OS entropy, so **the subset is redrawn on every process start**, and it is not recorded in the checkpoint.

That breaks resume, because the dataset is rebuilt *before* the checkpoint is loaded:

```python
# verl/trainer/ppo/v1/trainer_base.py
233:        self._init_dataloader()     # reconstructs RLHFDataset -> redraws the subset
369:        self._load_checkpoint()     # restores the sampler position into the new subset
```

and the restore only replays the sampler's index state:

```python
# verl/trainer/ppo/v1/trainer_base.py:836-839
dataloader_local_path = os.path.join(global_step_folder, "data.pt")
if os.path.exists(dataloader_local_path):
    dataloader_state_dict = torch.load(dataloader_local_path, weights_only=False)
    self.train_dataloader.load_state_dict(dataloader_state_dict)
```

The intent here is explicit — verl deliberately uses torchdata's stateful sampler for this:

```python
# verl/trainer/ppo/utils.py:154-156
# torch.utils.data.RandomSampler could not recover properly
from torchdata.stateful_dataloader.sampler import RandomSampler
# Use a sampler to facilitate checkpoint resumption.
```

Restoring an index position is only meaningful if the rows those indices address are the same. With an unseeded subset they are not.

#### Concrete numbers

`train.parquet` with 100,000 rows; `data.train_max_samples=10000`, `data.val_max_samples=500`, `data.train_batch_size=512`, `trainer.save_freq=10`, everything else default (`shuffle: True`, `seed: null`).

- Run 1 draws subset S1, trains steps 1-10, and checkpoints at step 10 with the sampler at index 5120.
- Restart with `resume_mode=auto`: the dataset draws a **different** subset S2, then the sampler position 5120 is restored into a permutation over S2.
- Expected overlap between two independent 10,000-of-100,000 draws is about 1,000 rows, i.e. roughly 90% of the "resumed" corpus is data the original run never saw, and roughly 90% of the original corpus is now unreachable.

Nothing warns: `len(dataset)` is unchanged, `steps_per_epoch` is unchanged, and training continues normally.

#### Second, related problem: the validation subset ignores `validation_shuffle`

Both datasets are constructed from the same `config.data` object:

```python
# verl/trainer/ppo/v1/trainer_base.py:659-674
self.train_dataset = create_rl_dataset(self.config.data.train_files, self.config.data, ...,
                                       max_samples=self.config.data.get("train_max_samples", -1))
self.val_dataset   = create_rl_dataset(self.config.data.val_files,   self.config.data, ...,
                                       max_samples=self.config.data.get("val_max_samples", -1))
```

so `RLHFDataset` reads `config.get("shuffle", False)` — the **train** flag — for the val subset too. `data.validation_shuffle: False` does not reach this code at all. Consequence: with `val_max_samples` set, the validation set is a different random 500 rows on every restart, so `val/*` metrics are not comparable across restarts of the same run.

#### Minimal check

No GPU needed:

```python
# two constructions of the same dataset with the same config select different rows
from verl.utils.dataset.rl_dataset import RLHFDataset
cfg = dict(shuffle=True, seed=None, filter_overlong_prompts=False, max_samples=10)
a = RLHFDataset(data_files="small.parquet", tokenizer=tok, config=cfg, max_samples=10)
b = RLHFDataset(data_files="small.parquet", tokenizer=tok, config=cfg, max_samples=10)
assert a.dataframe["extra_info"] == b.dataframe["extra_info"]   # currently fails
```

Setting `data.seed=42` makes both problems disappear, which confirms the diagnosis.

### Expected behavior

`max_samples` should select a reproducible subset so that a resumed run trains on the same data it was checkpointed on, and so that two runs with the same config are comparable.

Suggested fixes, in rough order of preference:

1. Make the draw deterministic by default — fall back to a fixed seed (or `trainer.seed`) instead of OS entropy when `data.seed` is unset. Smallest change; a resumed run then rebuilds the identical subset.
2. Persist the selected indices in the dataloader checkpoint (`data.pt`) and reuse them on resume. Robust even if the user changes `data.seed` between runs.
3. Independently of the above, have the validation dataset honour `data.validation_shuffle` rather than `data.shuffle`.

A CPU-only regression test is easy here — construct the dataset twice from a tiny parquet and assert the selected rows are stable.

I am happy to send a PR; option 1 (plus 3) looks like the minimal fix that restores the documented resume guarantee, but if maintainers prefer 2 I can do that instead.
